src/coreComponents/common/LifoStorage.hpp

```diff
@@ -0,0 +1,198 @@
+ * ------------------------------------------------------------------------------------------------------------
+ */
+#ifndef LIFOSTORAGE_HPP
+#define LIFOSTORAGE_HPP
+
+#include <deque>
+#include <future>
+#include <mutex>
+#include <condition_variable>
+#include <camp/camp.hpp>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <algorithm>
+
+#include "common/GEOS_RAJA_Interface.hpp"
+#include "common/TimingMacros.hpp"
+#include "common/LifoStorageCommon.hpp"
+#include "common/LifoStorageHost.hpp"
+#ifdef GEOSX_USE_CUDA
+#include "common/LifoStorageCuda.hpp"
+#endif
+
+namespace geosx
+{
+/**
+ * This class is used to store in a LIFO way buffers, first on device, then on host, then on disk.
+ */
+template< typename T, typename INDEX_TYPE >
+class LifoStorage
+{
+
+private:
+  /// number of buffers to be inserted into the LIFO
+  int m_maxNumberOfBuffers;
+  /// size of one buffer in bytes
+  size_t m_bufferSize;
+  /// counter of buffer stored in LIFO
+  int m_bufferCount;
+
+  /// pointer either to CUDA aware or host only LifoStorage
+  std::unique_ptr< LifoStorageCommon< T, INDEX_TYPE > > m_lifo;
+public:
+
+
+  /**
+   * A LIFO storage will store numberOfBuffersToStoreDevice buffer on
+   * deevice, numberOfBuffersToStoreHost on host and the rest on disk.
+   *
+   * @param name                           Prefix of the files used to save the occurenncy of the saved buffer on disk.
+   * @param elemCnt                        Number of elments in the LvArray we want to store in the LIFO storage.
+   * @param numberOfBuffersToStoreOnDevice Maximum number of array to store on device memory ( -1 = use 80% of remaining memory ).
+   * @param numberOfBuffersToStoreOnHost   Maximum number of array to store on host memory ( -1 = use 80% of remaining memory ).
+   * @param maxNumberOfBuffers             Number of arrays expected to be stores in the LIFO.
+   */
+  LifoStorage( std::string name, size_t elemCnt, int numberOfBuffersToStoreOnDevice, int numberOfBuffersToStoreOnHost, int maxNumberOfBuffers ):
+    m_maxNumberOfBuffers( maxNumberOfBuffers ),
+    m_bufferSize( elemCnt*sizeof( T ) ),
+    m_bufferCount( 0 )
+  {
+    LIFO_LOG_RANK( " LIFO : maximum size "<< m_maxNumberOfBuffers << " buffers " );
+    double bufferSize = ( ( double ) m_bufferSize ) / ( 1024.0 * 1024.0 );
+    LIFO_LOG_RANK( " LIFO : buffer size "<< bufferSize << "MB" );
+#ifndef GEOSX_USE_CUDA
+    numberOfBuffersToStoreOnDevice = 0;
+#else
+    if( numberOfBuffersToStoreOnDevice == -1 )
+    {
+      size_t free, total;
+      GEOSX_ERROR_IF( cudaSuccess != cudaMemGetInfo( &free, &total ), "Error getting CUDA device available memory" );
+      double freeGB = ( ( double ) free ) / ( 1024.0 * 1024.0 * 1024.0 );
+      LIFO_LOG_RANK( " LIFO : available memory on device " << freeGB << " GB" );
+      numberOfBuffersToStoreOnDevice = std::min( (int)( 0.8 * free / m_bufferSize ), m_maxNumberOfBuffers );
+    }
+#endif
```

Comment:
I think we want to remove this kind of
```c++
#ifndef GEOSX_USE_CUDA
    ...
#else
...
#endif
```
patterns in our code.
Is there a strong limitation that prevents you from putting
```
    numberOfBuffersToStoreOnDevice = 0;
```
in the `cuda` implementation and
```c++
    if( numberOfBuffersToStoreOnDevice == -1 )
    {
      size_t free, total;
      GEOSX_ERROR_IF( cudaSuccess != cudaMemGetInfo( &free, &total ), "Error getting CUDA device available memory" );
      double freeGB = ( ( double ) free ) / ( 1024.0 * 1024.0 * 1024.0 );
      LIFO_LOG_RANK( " LIFO : available memory on device " << freeGB << " GB" );
      numberOfBuffersToStoreOnDevice = std::min( (int)( 0.8 * free / m_bufferSize ), m_maxNumberOfBuffers );
    }
```
in the `host` one, and so on?
Eventually this file should only contain something like
```c++
#include "common/LifoStorageHost.hpp"
#ifdef GEOSX_USE_CUDA
#include "common/LifoStorageCuda.hpp"
#endif

#ifndef GEOSX_USE_CUDA
template< typename T, typename INDEX_TYPE >
class LifoStorage = LifoStorageCuda< T, INDEX_TYPE >
#else
template< typename T, typename INDEX_TYPE >
class LifoStorage = LifoStorageHost< T, INDEX_TYPE >
#endif
```

Reply:
In fact it's ifndef, I could remove the "numberOfBuffersToStoreOnDevice = 0;" which is not used. I didn't know how to put the numberOfBuffersToStoreOnDevice computation in the device part because it is used to initialize the FixedSizeDeque before the constructor content.
Maybe I can have a static method in each that compute it.

Reply:
_Without looking in the details, so I may be off topic:_
When you're in the `host` context, you know that `numberOfBuffersToStoreOnDevice` will be `-1` (IIUC).
Then you could directly provide the `-1` value for the initialization, without using any variable? (or something in the same mood).

Reply:
I cannot do exactly the using thing because in the case of 0 buffer on device, even if CUDA is available, I can't use the CUDA Lifo.
A solution to limit the number of ifdef would be to define LIfoStorageCuda pointing to a LifoCudaHost in case CUDA is not available.
Like that :
```
#include "common/LifoStorageHost.hpp"
#ifdef GEOSX_USE_CUDA
#include "common/LifoStorageCuda.hpp"
#else
template< typename T, typename INDEX_TYPE >
using LifoStorageCuda = LifoStorageHost< T, INDEX_TYPE >;
#endif
```
Not sure it helps readability...

Reply:
I proposed something tell me if that suit you.

Reply:
It seems the first solution does not work without CUDA, I changed it with a new one. Any comment @TotoGaz ?